GEM unpacker: bugfix and more cross checks

jshlee opened the pull request:
#### PR description:
* GEMDigiToRawModule issue with packing large data
  * Between 11_2_0pre2 to pre7, in phase2 PU samples, GEMMuons were not present which increased the fake rate in muons (https://indico.cern.ch/event/965774/#5-muo-gem-validation-studies).
  * This was found to be due to GEM digis not being produced (https://indico.cern.ch/event/965774/#6-gem-fed-issue).
  * GEM simdigis were not being packed correctly because only 12 AMCs can be stored per AMC13 (https://github.com/cms-sw/cmssw/blob/master/DataFormats/GEMDigi/interface/AMC13Event.h#L25), and with multiple BXs, PU events made many more AMCs which couldn't be unpacked. 
  * added more cross checks

* GEMRawToDigiModule
  * GEM digi bx is wrong, it should be the relative difference between AMC and BFAT bx.
  * crash protection from bad GEM chamber ID from DAQ

* No changes are expected for non PU samples.
* PU should be back to 11_2_0pre1

#### PR validation:
* tested with wf 11611.0, 23211.0 and run 337972 MWGR data.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
* backport might be needed for the crash protection

jshlee committed:
GEMDigiToRawModule - fix AMC size in packing
only 12 AMCs can be readout
GEMRawToDigiModule - fix gem digi bx

jshlee committed:
remove commented code

jshlee committed:
added more sanity checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31863/19221

- This PR adds an extra 28KB to repository

cmsbuild commented:
A new Pull Request was created by  @jshlee (Jason Lee) for master.

It involves the following packages:

DataFormats/GEMDigi
EventFilter/GEMRawToDigi


@perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77, @jpata can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @dildick, @rovere, @watson-ij this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

kpedro88 commented on EventFilter/GEMRawToDigi/plugins/GEMRawToDigiModule.cc:
```diff
@@ -118,25 +119,69 @@ void GEMRawToDigiModule::produce(edm::StreamID iID, edm::Event& iEvent, edm::Eve
+    }
+    // compare trailers
+    if (amc13Event->fragmentLength() != trailer.fragmentLength()) {
+      edm::LogWarning("GEMRawToDigiModule") << " fragmentLength miss match";
```
typo: miss match -> mismatch (also in next few lines)

kpedro88 replied:
in all of these messages (also in the next file), it might be useful to print the actual mismatched values as well

jshlee replied:
thanks @kpedro88! 
I'll fix them after the test

jshlee replied:
The values for amc13Event and trailer would not be that useful, since they should be reading the same data.
This test is to just make sure that they are both reading the last data line.

kpedro88 commented on EventFilter/GEMRawToDigi/src/GEMRawToDigi.cc:
```diff
@@ -38,16 +45,22 @@ std::unique_ptr<AMC13Event> GEMRawToDigi::convertWordToAMC13Event(const uint64_t
-        return nullptr;
+      if (gebData.vfatWordCnt() != gebData.vfatWordCntT()) {
+        vfatError_ = true;
+        edm::LogWarning("GEMRawToDigi") << "VFAT word count miss match between header and trailer";
```
mismatch

jshlee replied:
done

kpedro88 commented on EventFilter/GEMRawToDigi/src/GEMRawToDigi.cc:
```diff
@@ -38,16 +45,22 @@ std::unique_ptr<AMC13Event> GEMRawToDigi::convertWordToAMC13Event(const uint64_t
+                             << " amc13Event->getAMCsize(i): " << int(amc13Event->getAMCsize(i));
+    if (amc13Event->getAMCsize(i) != amcData.dataLength()) {
+      amcError_ = true;
+      edm::LogWarning("GEMRawToDigi") << "AMC size miss match";
```
mismatch

jshlee replied:
done

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d7cec9eb41ec635e29670a6838f4b5983d0d16a2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10103/summary.html
CMSSW: CMSSW_11_2_X_2020-10-19-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10103/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 9 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2543752
 - DQMHistoTests: Total failures: 19
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2543711
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jshlee committed:
typo

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31863/19231

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #31863 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77, @jpata can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ff40bab0807069026917f2b82f528e57b1d3d0da
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10110/summary.html
CMSSW: CMSSW_11_2_X_2020-10-19-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10110/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 13 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2543752
 - DQMHistoTests: Total failures: 22
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2543708
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

civanch commented:
+1

slava77 commented:
I see quite a few warnings added in this PR. Some apparently can happen as deep as per data frame in the AMC payloads.
What are the expectations for the frequency of these?
On average these better stay at o(1e-3) and on problematic events we better have just one or two messages.
I'd like to a void a flood of messages, which can trigger problems for T0/production.

@jshlee 
please check and clarify.

Also, will this need a backport to 11_1_X?

jshlee commented:
@slava77 It's a bit hard to estimate the frequency of these messages in data. 
On non problematic events, we should get no outputs. 
On problematic events, I imagine quite a few messages popping up at the same time.
I can reduce these to somewhat of a condensed summary warning.

Looks like we will need a backport to 11_1_X.

slava77 commented:
>  I can reduce these to somewhat of a condensed summary warning.

a summary would be nice (per FED or even just per event).

BTW, is the data with warnings still usable or simply not written to the digis? I wanted to make sure that the consumers downstream get usable data and would not need to add checks/patches.

jshlee commented:
will reduce to 1 warning per FED

Currently, it produces digis if it can, so in some cases it can make digis, but I wouldn't trust the data from it.
I'll fix to not producing digis if warnings come up.

slava77 commented:
> Currently, it produces digis if it can, so in some cases it can make digis, but I wouldn't trust the data from it.
> I'll fix to not producing digis if warnings come up.

if the data is still needed for some analysis of issues, it may be practical to have a data product with problematic digis (either verbatim or with some tag/quality data added).

jshlee committed:
reduced warnings to 1 per FED

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31863/19267

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #31863 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77, @jpata can you please check and sign again.

jshlee commented:
> > Currently, it produces digis if it can, so in some cases it can make digis, but I wouldn't trust the data from it.
> > I'll fix to not producing digis if warnings come up.
> 
> if the data is still needed for some analysis of issues, it may be practical to have a data product with problematic digis (either verbatim or with some tag/quality data added).

I've reduced the warnings to 1 per FED.
For the problematic digis, I broke it into two main cases.
If the FED isn't read properly, the CDF trailers will end up different and I think it's best to reject these events since the words can get mixed up.
If the FED is read properly, but looks for GEBs and VFATs which are not in the electronics map,only these GEBs and VFATs are skipped, while the other digis are produced. I think it would be good to add a flag for this to show up in onlineDQM to help us debug this issue.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on EventFilter/GEMRawToDigi/interface/GEMRawToDigi.h:
```diff
@@ -9,7 +9,11 @@
 class GEMRawToDigi {
 public:
   std::unique_ptr<gem::AMC13Event> convertWordToAMC13Event(const uint64_t* word);
+  bool vfatError() const { return vfatError_; }
+  bool amcError() const { return amcError_; }
```
I do not see use cases.
Are these added for some future use or was some code accidentally not committed?

jshlee replied:
It was used, but I removed the use cases when I reduced the LogWarnings. 
I would like to add a few of these checks to the onlineDQM.

cmsbuild commented:
+1
Tested at: d89b13e838ed72bed7ad15cf53fb757823277ea7
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10160/summary.html
CMSSW: CMSSW_11_2_X_2020-10-20-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10160/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544100
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544069
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jshlee committed:
added back edm::Log category strings

jshlee commented on EventFilter/GEMRawToDigi/plugins/GEMRawToDigiModule.cc:
```diff
@@ -220,6 +258,11 @@ void GEMRawToDigiModule::produce(edm::StreamID iID, edm::Event& iEvent, edm::Eve
     }
 
+    if (unknownChamber || unknownVFat || badVfat) {
+      edm::LogWarning("") << "unpacking error: unknown Chamber " << unknownChamber << " unknown VFat " << unknownVFat
```
I removed them since the outputs always contain the same information without it.
I put them back in

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31863/19277

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #31863 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77, @jpata can you please check and sign again.

perrotta commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: deaa6455dc7c7e1f754a787ceceb77b952efac1a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10169/summary.html
CMSSW: CMSSW_11_2_X_2020-10-20-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10169/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544100
 - DQMHistoTests: Total failures: 15
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544063
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

slava77 commented:
I'm not sure about the updated code in this PR based of DQM plots:

in no-PU sample we see some digis disappearing (wf 10024, ttbar no-PU 2017)
<img width="740" alt="wf10024 0_gem_strip_bx" src="https://user-images.githubusercontent.com/4676718/96735587-901cbf00-1370-11eb-8b2c-4258272cca2f.png">
From the PR description I somehow understood that things should be mostly unchanged.

I looked at the reconstructed hits `GEMDetIdGEMRecHitsOwnedRangeMap_gemRecHits__RECO.obj.collection_.data_.BunchX()`
in the baseline we have 2 hits with BX values 1 and 0, while with the PR the BX values are 0 and -4095, respectively. I somehow though that with the PR the BX values should be more meaningful.

It's less clear what's in the sample with PU (wf 10224, ttbar with PU 2017)
<img width="741" alt="wf10224 0_gem_strip_bx" src="https://user-images.githubusercontent.com/4676718/96735841-d540f100-1370-11eb-8f96-e0d20da71d15.png">

at the hit level the BX values are shifting as well in a way that I didn't quite expect
```
      row  * inst   *   new BX  *   old BX *
*        0 *        0 *       -12 *      4084 *
*        0 *        1 *       -12 *      4084 *
*        1 *        0 *         0 *         0 *
*        1 *        1 *         0 *         0 *
*        2 *        0 *       -12 *      4084 *
*        2 *        1 *     -4095 *         1 *
*        2 *        2 *       -12 *      4084 *
*        2 *        3 *         0 *         0 *
*        3 *        0 *       -12 *       -12 *
*        4 *        0 *        -8 *      4088 *
*        4 *        1 *        -8 *      4088 *
*        5 *        0 *        -1 *      4095 *
*        7 *        0 *     -4094 *         2 *
*        7 *        1 *     -4094 *         2 *
*        8 *        0 *     -4095 *         1 *
*        8 *        1 *     -4095 *         1 *
```
the new values at 4095 look suspicious (although so many -12 is a bit odd as well, but at least mixing includes a valid BX at this value)

jshlee commented:
@slava77 I'll test these workflows

slava77 commented:
> @slava77 I'll test these workflows

thank you.

Now I looked at my test sample in the 2021 setup; the sample  has "inverted" pileup: minBias is the signal and a flat muon gun is used as pileup with average pileup of 10; this scans the issue well.

In the baseline things are simple: there are no hits at all; with the PR the hits appear and their BX is in the acceptable range [-13 , 4] out of 30K hits total.
the plots also look encouraging

<img width="743" alt="wf11840mix-mu_gemHits_softMu" src="https://user-images.githubusercontent.com/4676718/96739519-b8a6b800-1374-11eb-99c2-46253d659a27.png">
<img width="740" alt="wf11840mix-mu_gem_sta_eff" src="https://user-images.githubusercontent.com/4676718/96739536-be040280-1374-11eb-877a-81a16d6492c8.png">
<img width="739" alt="wf11840mix-mu_gem_strip_bx" src="https://user-images.githubusercontent.com/4676718/96739552-c1978980-1374-11eb-8b6e-8029bd2bc428.png">

So, I suspect something is special about the 2017 (and probably 2018) case. 
Are they using hardcoded calibrations?
... didn't the problem show up after we switched to the GEM emap from DB?

jshlee commented:
@slava77 this is due to the different dataformat in VFAT for 2017. https://github.com/cms-sw/cmssw/blob/master/DataFormats/GEMDigi/interface/VFATdata.h#L29
VFAT v2 is used, which only has 12 bits for vfat.bc. 
AMC.bc is also 12 bits.

      row  * inst   *   new BX  *   old BX *
*        0 *        0 *       -12 *      4084 *
*        0 *        1 *       -12 *      4084 *
*        1 *        0 *         0 *         0 *
*        1 *        1 *         0 *         0 *
*        2 *        0 *       -12 *      4084 *
*        2 *        1 *     -4095 *         1 *

in the last line, I suspect AMC.bc = 4095 and vfat.bc = 4096, but since vfat.bc is only 12bits, vfat.bc ends up being 0.
So we end up with bx = 0-4095.

slava77 commented:
> @slava77 this is due to the different dataformat in VFAT for 2017.

I guess this explains how -4095 appears, but I couldn't understand if it's the needed value.
does it mean that this should be fixed in this PR?

jshlee commented:
> > @slava77 this is due to the different dataformat in VFAT for 2017.
> 
> I guess this explains how -4095 appears, but I couldn't understand if it's the needed value.
> does it mean that this should be fixed in this PR?

No, I didn't expect this to happen. I will look for a fix.

jshlee committed:
fix bx issues for mc using vfat v2

cmsbuild commented:
The code-checks are being triggered in jenkins.

jshlee commented:
@slava77 I've fixed this issue. The issue was that for MC, the Event.bunchCrossing() was always -1, which is not nice to save as 12 bits. I fixed by converting Event.bunchCrossing() to uint8_t.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31863/19292

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #31863 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77, @jpata can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on EventFilter/GEMRawToDigi/plugins/GEMDigiToRawModule.cc:
```diff
@@ -130,7 +130,6 @@ void GEMDigiToRawModule::produce(edm::StreamID iID, edm::Event& iEvent, edm::Eve
     for (uint8_t amcNum = 0; amcNum < GEMeMap::maxAMCs_; ++amcNum) {
       uint32_t amcSize = 0;
       std::unique_ptr<AMCdata> amcData = std::make_unique<AMCdata>();
-      amcData->setAMCheader1(amcSize, BX_id, LV1_id, amcNum);
```
is this header assignment part really not needed?

jshlee replied:
yes, it is set later on anyways.
I was mainly using it for the 16bit to 12bit conversion of BX_id

slava77 replied:
IIUC, it will be set only if `!gebData->vFATs()->empty()`.
I suppose a real AMC without any VFATs will still send a header.
Perhaps the point for MC is that it doesn't matter.

jshlee replied:
Yes, for MC it doesn't matter.
I can remove the if !gebData->vFATs()->empty() if you prefer it.

slava77 replied:
naively, a better representation of reality would make more sense. E.g. with the headers/trailers properly set the unpacker would be more capable of doing integrity checks without some kind of `if(realData)`. But perhaps it's an academic topic and it doesn't really matter.

I don't have a strong preference.

jshlee replied:
I've changed to always producing AMCs.

cmsbuild commented:
+1
Tested at: ef24cdce1496ee59f63c23ecff1ae666b4da7876
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10186/summary.html
CMSSW: CMSSW_11_2_X_2020-10-21-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10186/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 13 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544110
 - DQMHistoTests: Total failures: 39
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544049
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

slava77 commented on EventFilter/GEMRawToDigi/plugins/GEMDigiToRawModule.cc:
```diff
@@ -123,37 +123,36 @@ void GEMDigiToRawModule::produce(edm::StreamID iID, edm::Event& iEvent, edm::Eve
-            GEMDetId gemId = vfat_dc.detId;
-            uint16_t vfatId = vfat_ec.vfatAdd;
+          for (auto const& gemBx : gemBxMap) {
+            int bc = amcData->bx() + gemBx.first;
```
actually that header we talked about in https://github.com/cms-sw/cmssw/pull/31863#discussion_r509607308 is effectively used here. So, instead of 255 (uint8_t(-1)) we are getting 0 while packing. But then during unpacking the AMC BX is 255, which leads to the bad shift observed in the tests.

jshlee replied:
this is fix. it should have been 
```
int bc = BX_id + gemBx.first;
```

slava77 commented:
it looks like the latest updates opened a cycle for another update: all strip and rechit BX is shifted down by -255 from their "normal" range ([-13 , 4]  in my mixed muons sample).

silviodonato commented:
urgent
@jshlee we are going to close pre8 this night (CERN time)

jshlee committed:
GEMDigiToRawModule - always make AMCs

cmsbuild commented:
The code-checks are being triggered in jenkins.

jshlee committed:
code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

jshlee commented:
> urgent
> @jshlee we are going to close pre8 this night (CERN time)

I think all issues so far have been solved.

jshlee commented:
> it looks like the latest updates opened a cycle for another update: all strip and rechit BX is shifted down by -255 from their "normal" range ([-13 , 4] in my mixed muons sample).

this is also due to using amcData->bx() instead of BX_id

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31863/19322

- This PR adds an extra 40KB to repository

cmsbuild commented:
Pull request #31863 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @jpata, @mdhildreth, @slava77, @ggovi can you please check and sign again.

silviodonato commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7433f4cbeac77d60188758fca5f1d8fa3d2a2973
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10211/summary.html
CMSSW: CMSSW_11_2_X_2020-10-21-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10211/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544110
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544081
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

slava77 commented on CondFormats/GEMObjects/interface/GEMeMap.h:
```diff
@@ -19,7 +19,7 @@ class GEMeMap {
 
   struct GEMChamberMap {
-    std::vector<unsigned int> fedId;
+    std::vector<int> fedId;
```
I hope this is a transparent change for streaming. I recall that the conditions formats are not really modifiable in an easy way (no schema evolution)

jshlee replied:
i will just revert back

slava77 commented on DataFormats/GEMDigi/interface/AMCdata.h:
```diff
@@ -92,7 +92,7 @@ namespace gem {
 
     uint16_t boardId() const { return AMCheader2{amch2_}.boardID; }
-    uint16_t orbitNum() const { return AMCheader2{amch2_}.orbitNum; }
+    int16_t orbitNum() const { return AMCheader2{amch2_}.orbitNum; }
```
everything else is uint\*; is orbitNum special?

jshlee replied:
not special, just the default MC value is -1

slava77 replied:
but so is `bunchCrossing()`; to be consistent, shouldn't the bx data member be changed as well?

jshlee replied:
I reverted back for now. 
I changed orbitNum because it was 16 bits while bunchCrossing is only 12 bits.

slava77 commented on EventFilter/GEMRawToDigi/src/GEMRawToDigi.cc:
```diff
@@ -3,7 +3,8 @@
 #include "EventFilter/GEMRawToDigi/interface/GEMRawToDigi.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
-
+// temp
+#include <iostream>
```
this looks like a leftover from some debugging session.

jshlee replied:
yep. will remove

slava77 reviewed (commented):
it looks like other than the proper fix, a few more unrelated items were added.
Were they really needed now?

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/31863	7433f4c 
- code changes are in line with the PR description and the follow up review; the reco part looks OK now (ignoring the extra `// temp` in https://github.com/cms-sw/cmssw/pull/31863#discussion_r510300043); a few other changes in the last commits look a bit odd, but they are not in the reco category
- jenkins tests pass and comparisons with the baseline show differences in GEM BX ID; a manual check of jenkins outputs for wf 10224 shows that the rechits with this PR are now in the normal range of BX; the slightly older tests in a muon-mixed workflow should still apply https://github.com/cms-sw/cmssw/pull/31863#issuecomment-713647780 and confirm the primary goal of the bugfix is achieved.

jshlee committed:
code cleanup

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31863/19330

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #31863 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @jpata, @mdhildreth, @slava77, @ggovi can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

silviodonato commented:
@cms-sw/simulation-l2 @cms-sw/upgrade-l2 @cms-sw/db-l2 
do you have any comments?
I would try to merge this before the next IB to include it in pre8 (this is a long pending issue discussed at the PPD meetings, it would be good to get it fixed in the next round of validation)

kpedro88 commented:
I have no other comments; will sign once tests pass.

cmsbuild commented:
-1

Tested at: 78a7536d48642cf02c37db8d6965c47d65d53d53

CMSSW: CMSSW_11_2_X_2020-10-21-2300
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10217/summary.html

I found follow errors while testing this PR

**Failed tests:  AddOn**

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py RelVal -s HLT:PIon,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run3_mc_PIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3 --processName=HLTRECO --filein file:RelVal_Raw_PIon_MC.root --fileout file:RelVal_Raw_PIon_MC_HLT_RECO.root : FAILED - time: date Thu Oct 22 20:07:34 2020-date Thu Oct 22 19:59:32 2020 s - exit: 34304

cmsbuild commented:
Comparison job queued.

silviodonato commented:
> -1
> 
> Tested at: [78a7536](https://github.com/cms-sw/cmssw/commit/78a7536d48642cf02c37db8d6965c47d65d53d53)
> 
> CMSSW: CMSSW_11_2_X_2020-10-21-2300
> SCRAM_ARCH: slc7_amd64_gcc820
> You can see the results of the tests here:
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10217/summary.html
> 
> I found follow errors while testing this PR
> 
> **Failed tests: AddOn**
> 
>     * **AddOn**:
> 
> 
> I found errors in the following addon tests:
> 
> cmsDriver.py RelVal -s HLT:PIon,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run3_mc_PIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3 --processName=HLTRECO --filein file:RelVal_Raw_PIon_MC.root --fileout file:RelVal_Raw_PIon_MC_HLT_RECO.root : FAILED - time: date Thu Oct 22 20:07:34 2020-date Thu Oct 22 19:59:32 2020 s - exit: 34304
The error is in the very last line of
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10217/addOnTests/hlt_mc_PIon/cmsDriver.py_RelVal_-s_HLT:PIon,RAW2DIGI,L1Reco,RECO_--mc_--scenario=pp_-n_10_--conditions_auto:run3_mc_PIon_--relval_9000,50_--datatier_RAW-HLT-RECO.log

`Fatal system signal has occurred during exit`

I hope it is just a transient problem

silviodonato commented:
Yes, we saw that same problem here https://github.com/cms-sw/cmssw/pull/31602#issuecomment-703127865 and it went away by running one more time the test.

Same thing here https://github.com/cms-sw/cmssw/pull/31792#issuecomment-708910667

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/31863	78a7536 
- most recent updates since the last reco signoff were technical https://github.com/cms-sw/cmssw/pull/31863/files/7433f4cbeac77d60188758fca5f1d8fa3d2a2973..78a7536d48642cf02c37db8d6965c47d65d53d53

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa155c/10217/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544110
 - DQMHistoTests: Total failures: 16
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544072
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

silviodonato commented:
merge
